flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java

```diff
@@ -706,8 +706,10 @@ private Execution resetForNewExecutionInternal(final long timestamp, final long
 
 	public void tryAssignResource(LogicalSlot slot) {
 		if (!currentExecution.tryAssignResource(slot)) {
-			throw new IllegalStateException("Could not assign resource " + slot + " to current execution " +
-				currentExecution + '.');
+			final IllegalStateException e = new IllegalStateException(
+				"Could not assign resource " + slot + " to current execution " + currentExecution + '.');
+			slot.releaseSlot(e);
```

Comment:
I am getting second thoughts about this change. It's not optimal that we are adding business logic to the `ExecutionVertex`. I think it's better to let the caller handle releasing the slot. Since this `IllegalStateException` denotes a bug, what do you think about wrapping all instances of `.handle(assignResourceOrHandleError(deploymentHandle))` into `FutureUtils.assertNoException`?

Reply:
Ok. Agreed that `tryAssignResource` should never fail in `DefaultScheduler#assignResourceOrHandleError` unless there is a bug.
Let's drop this hotfix.

I think we do not need to change `assertNoException`. Changing `.handle(deployOrHandleError...` to `.thenAccept(deployOrHandleError...` should work for this purpose. 
And I'd prefer to do this after we have unified the deployment logic in FLINK-14162.

Reply:
Actually I can see that in `waitForAllSlotsAndDeploy` such unexpected errors are already propagated from `assignResourceOrHandleError` to `deployAll`, then to `assertNoException`

Reply:
I mean the `deployAll` process.
`deployIndividually` has the issue you mentioned. But I think we will remove this code path in FLINK-14162.

Reply:
You are right.